src/composables/Alchemy.ts

```diff
@@ -317,3 +335,85 @@ export const VialCost = [
   100e6,
   1e9,
 ];
+
+export class Alch {
+  static discount (cauldCostReduxLvl: number, bubbleCostBubbleLvl: number, bubbleCostVialLvl: number, bubbleTwelveLvl: number, tagLvl: number): any {
+    const precision = 10000;
+
+    const costReduxBoost = Math.round(10 * Growth["Decay"](cauldCostReduxLvl, 90, 100)) / 10;
+    const oa          = Math.max(0.1, 1 - costReduxBoost / 100); // TODO: This one is off, but the total calculation is still correct
+    const newBubble   = Math.max(0.05, 1 - (Growth['Decay'](bubbleTwelveLvl, 40, 12)/100)); // Hardcoded values, better to retrieve from bubbles data maybe.
+    const undevCost   = Growth["Decay"](bubbleCostBubbleLvl, 40, 70);
+    const vial        = Growth["Add"](bubbleCostVialLvl, 1, 0);
+    const undev_vial  = Math.max(0.05, 1 - (undevCost + vial) / 100);
+    const bargain_tag = Math.max(Math.pow(0.75, tagLvl), 0.1);
+    var discount = oa * newBubble * undev_vial * bargain_tag;
+    var result = [oa, bargain_tag, newBubble, undev_vial, discount];
+    result = result.map((a) => {
+        return (precision - Math.round(a*precision))/100}
+      );
+
+    console.log(`Alch.discount = [
+      Cauldron:     ${result[0].toFixed(2).padStart(5, " ")} 
+      Bargain:      ${result[1].toFixed(2).padStart(5, " ")} 
+      Bubble XII:   ${result[2].toFixed(2).padStart(5, " ")} 
+      Undev + vial: ${result[3].toFixed(2).padStart(5, " ")} 
+      Total:        ${result[4].toFixed(2).padStart(5, " ")}
+    ]`);
+    return result;
+  }
+  
+  static effect = (bubble: Bubble, level: number) => {
+    let funcStr = bubble.Func;
+    let funcStrCap = funcStr.charAt(0).toUpperCase() + funcStr.slice(1);
+
+    let effect = (level === 0 ? 0 : Growth[funcStrCap](level, Number(bubble.x1), Number(bubble.x2)));
+    if (isNaN(effect)) {
+      effect = 0;
+    }
+    return Number(effect);
+  }
+
+  static effectChange = (bubble: Bubble, levelNow: number, levelGoal: number) => {
+    let effectNow = Alch.effect(bubble, levelNow) ?? 0;
+    let effectGoal = Alch.effect(bubble, levelGoal);
+    effectGoal = (effectGoal < effectNow ? effectNow : effectGoal);
+    let result = `${Number(effectNow).toFixed(2).padStart(6, ' ')} => ${Number(effectGoal).toFixed(2).padStart(6, ' ')}`;
+    return result;
+  };
+
+
+  static Pay2Win = (levelNow: any, levelWant: any, type: any) => {
+    let totalSum = 0;
+    let result = 0;
+    for (var j = levelNow; j < levelWant; j++) {
+      switch(type) { 
+        case 0: // Speed
```

Comment:
Instead of these comments, use an enum in these cases:
```ts
enum AlchemyPay2Win {
  Speed,
  NewBubble,
  BoostReq,
}

// Access as:
AlchemyPay2Win.BoostReq // === 2
```